Parse reflection info after linking programs

jrprice opened the pull request:
Otherwise push constants, spec constants, and required work-group size information will not be present in the linked binary.

This fixes some CTS regressions introduced by #283. The Clspv update is needed for a recent bugfix that affects the test case added in this PR.

kpet commented on src/program.cpp:
```diff
@@ -896,9 +896,10 @@ cl_build_status cvk_program::link() {
 
-    // Merge descriptor maps
-    for (cl_uint i = 0; i < m_num_input_programs; i++) {
-        m_binary.insert_descriptor_map(m_input_programs[i]->m_binary);
```
This was the last use of `spir_binary::insert_descriptor_map`. We can delete the function entirely as well as `spir_binary::m_dmaps_text`. Good riddance :).

kpet commented on tests/api/compiler.cpp:
```diff
@@ -35,3 +35,61 @@ TEST_F(WithContext, BuildLog) {
     ASSERT_TRUE(build_log.find("THIS IS AN ERROR") != std::string::npos);
 }
+
+TEST_F(WithCommandQueue, SimpleLink) {
```
Could you add a short comment describing what is being tested please? Maybe a test name change is good enough. `CompileAndLinkWithPushConstants`?

kpet commented on tests/api/compiler.cpp:
```diff
@@ -35,3 +35,61 @@ TEST_F(WithContext, BuildLog) {
+    // Enable 2.0 to generate push constants for the global size.
+    err = clCompileProgram(program, 1, &gDevice, "-cl-std=CL2.0", 0, nullptr,
+                           nullptr, nullptr, nullptr);
+    ASSERT_EQ(err, CL_SUCCESS);
```
Please add a `CompileProgram` helper.

kpet commented on tests/api/compiler.cpp:
```diff
@@ -35,3 +35,61 @@ TEST_F(WithContext, BuildLog) {
+    holder<cl_program> linked_program =
+        clLinkProgram(m_context, 1, &gDevice, nullptr, 1, &program_list,
+                      nullptr, nullptr, &err);
+    ASSERT_EQ(err, CL_SUCCESS);
```
Same here, please add a helper.

kpet commented on tests/api/compiler.cpp:
```diff
@@ -35,3 +35,61 @@ TEST_F(WithContext, BuildLog) {
+    auto gws_output = CreateBuffer(CL_MEM_READ_WRITE, sizeof(cl_uint), nullptr);
+    auto lws_output = CreateBuffer(CL_MEM_READ_WRITE, sizeof(cl_uint), nullptr);
+    auto dim_output = CreateBuffer(CL_MEM_READ_WRITE, sizeof(cl_uint), nullptr);
+    cl_uint value = 42;
```
This seems unused.

kpet commented on tests/api/compiler.cpp:
```diff
@@ -35,3 +35,61 @@ TEST_F(WithContext, BuildLog) {
+
+    err = clEnqueueReadBuffer(m_queue, gws_output, CL_TRUE, 0, sizeof(cl_uint),
+                              &result, 0, nullptr, nullptr);
+    ASSERT_EQ(err, CL_SUCCESS);
```
Same here, please add a helper.

kpet commented on tests/api/compiler.cpp:
```diff
@@ -35,3 +35,61 @@ TEST_F(WithContext, BuildLog) {
+
+    auto gws_output = CreateBuffer(CL_MEM_READ_WRITE, sizeof(cl_uint), nullptr);
+    auto lws_output = CreateBuffer(CL_MEM_READ_WRITE, sizeof(cl_uint), nullptr);
+    auto dim_output = CreateBuffer(CL_MEM_READ_WRITE, sizeof(cl_uint), nullptr);
```
Might be worth adding a new overload for `CreateBuffer` with no host pointer or giving `host_ptr` a default argument.

jrprice replied:
I added the overload. For consistency I also did the same for `CreateImage`.

jrprice commented:
Thanks for the quick review! I've addressed the review comments. I also added another test to cover literal samplers, since they would be hit by the same issue.

kpet commented on tests/api/compiler.cpp:
```diff
@@ -47,24 +48,17 @@ TEST_F(WithCommandQueue, SimpleLink) {
-        clLinkProgram(m_context, 1, &gDevice, nullptr, 1, &program_list,
-                      nullptr, nullptr, &err);
-    ASSERT_EQ(err, CL_SUCCESS);
+    holder<cl_program> linked_program = LinkProgram(1, &program_list);
```
Nit: could use `auto` to declare `linked_program`.

kpet commented on tests/api/compiler.cpp:
```diff
@@ -76,20 +70,88 @@ TEST_F(WithCommandQueue, SimpleLink) {
-    err = clEnqueueReadBuffer(m_queue, gws_output, CL_TRUE, 0, sizeof(cl_uint),
-                              &result, 0, nullptr, nullptr);
-    ASSERT_EQ(err, CL_SUCCESS);
+    EnqueueReadBuffer(gws_output, CL_TRUE, 0, sizeof(cl_uint), &result);
```
Nit: `CL_TRUE` could be `CL_BLOCKING`. Same for the two other calls.

kpet reviewed (approved):
LGTM, thanks!

jrprice commented:
Addressed nits, PTAL.

kpet reviewed (approved):
Ta!
